cmd/snap/cmd_run.go

```diff
@@ -1013,3 +1062,256 @@ func (x *cmdRun) runSnapConfine(info *snap.Info, securityTag, snapApp, hook stri
+// - the presence of the environment variable DBUS_SESSION_BUS_ADDRESS
+// - the presence of the bus socket address in the file /run/user/UID/dbus-session
+// - the presence of the bus socket in /run/user/UID/bus
+func isSessionBusLikelyPresent() bool {
+	if address := os.Getenv("DBUS_SESSION_BUS_ADDRESS"); address != "" {
+		return true
+	}
+	if fi, err := os.Stat(fmt.Sprintf("/run/user/%d/dbus-session", osGetuid())); err == nil {
+		if stat, ok := fi.Sys().(*syscall.Stat_t); ok {
+			if stat.Mode&syscall.S_IFREG == syscall.S_IFREG {
+				return true
+			}
+		}
+	}
+	if fi, err := os.Stat(fmt.Sprintf("/run/user/%d/bus", osGetuid())); err == nil {
+		if stat, ok := fi.Sys().(*syscall.Stat_t); ok {
+			if stat.Mode&syscall.S_IFSOCK == syscall.S_IFSOCK {
+				return true
+			}
+		}
+	}
+	return false
+}
+
+var dbusSessionBus = func() (*dbus.Conn, error) {
+	// This started as a helper for mocking and testing but we ended up needing
+	// a custom implementation to prevent go-dbus from spawning and *leaving
+	// behind* a dbus-launch process that will never be used by anything again.
+	if isSessionBusLikelyPresent() {
+		return dbus.SessionBus()
+	}
+	return nil, fmt.Errorf("cannot find session bus")
+}
+
+var randomUUID = func() (string, error) {
+	// The source of the bytes generated here is the same as that of
+	// /dev/urandom which doesn't block and is sufficient for our purposes
+	// of avoiding clashing UUIDs that are needed for all of the non-service
+	// commands that are started with the help of this UUID.
+	uuidBytes, err := ioutil.ReadFile("/proc/sys/kernel/random/uuid")
+	return strings.TrimSpace(string(uuidBytes)), err
+}
+
+var osGetpid = os.Getpid
+
+var cgroupProcessPathInTrackingCgroup = cgroup.ProcessPathInTrackingCgroup
+
+var errCannotTrackProcess = errors.New("cannot track application process")
+
+var createTransientScope = func(securityTag string) error {
+	if !features.RefreshAppAwareness.IsEnabled() {
+		return nil
+	}
+	logger.Debugf("creating transient scope %s", securityTag)
+
+	// We ask the kernel for a random UUID. We need one because each transient
+	// scope needs a unique name. The unique name is comprosed of said UUID and
+	// the snap security tag.
+	uuid, err := randomUUID()
+	if err != nil {
+		return err
+	}
+
+	// Enforcing uniqueness is preferred to reusing an existing scope for
+	// simplicity since doing otherwise and joining an existing scope has
+	// limitations:
+	// - the originally started scope must be marked as a delegate, with all consequences.
+	// - the method AttachProcessesToUnit is unavailable on Ubuntu 16.04
+	//
+	// TODO: invert the arguments so that security tag comes first, UUID
+	// second. This will result in nicer scope names but needs to be
+	// coordinated with code in sandbox/cgroup and with the spread tests
+	// exercising that.
+	unitName := fmt.Sprintf("snap.%s.%s.scope", uuid, strings.TrimPrefix(securityTag, "snap."))
```

Comment:
this bit of code probably belongs to sandbox/cgroup given that it has code to parse it, otherwise the two bits feel disconnected

Reply:
Indeed and now moved to sandbox/cgroup